File: midolmanj-mgmt/src/main/java/com/midokura/midolman/mgmt/host/rest_api/TunnelZoneResource.java
/*
 * Copyright 2012 Midokura PTE LTD.
 */
package com.midokura.midolman.mgmt.host.rest_api;

import com.google.inject.Inject;
import com.google.inject.servlet.RequestScoped;
import com.midokura.midolman.mgmt.ResourceUriBuilder;
import com.midokura.midolman.mgmt.VendorMediaType;
import com.midokura.midolman.mgmt.auth.AuthRole;
import com.midokura.midolman.mgmt.host.TunnelZone;
import com.midokura.midolman.mgmt.host.TunnelZoneFactory;
import com.midokura.midolman.mgmt.rest_api.BadRequestHttpException;
import com.midokura.midolman.mgmt.rest_api.NotFoundHttpException;
import com.midokura.midolman.mgmt.rest_api.ResourceFactory;
import com.midokura.midolman.state.StateAccessException;
import com.midokura.midonet.cluster.DataClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.security.RolesAllowed;
import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@RequestScoped
public class TunnelZoneResource {

    private final static Logger log = LoggerFactory
            .getLogger(TunnelZoneResource.class);

    private final UriInfo uriInfo;
    private final DataClient dataClient;
    private final Validator validator;
    private final ResourceFactory factory;

    @Inject
    public TunnelZoneResource(UriInfo uriInfo, DataClient dataClient,
                              Validator validator, ResourceFactory factory) {
        this.uriInfo = uriInfo;
        this.dataClient = dataClient;
        this.validator = validator;
        this.factory = factory;
    }

    @GET
    @RolesAllowed({AuthRole.ADMIN})
    @Produces({VendorMediaType.APPLICATION_TUNNEL_ZONE_COLLECTION_JSON,
                  MediaType.APPLICATION_JSON})
    public List<TunnelZone> list() throws StateAccessException {

        List<com.midokura.midonet.cluster.data.TunnelZone<?, ?>>
                tunnelZoneDataList = dataClient.tunnelZonesGetAll();
        List<TunnelZone> tunnelZones = new ArrayList<TunnelZone>();
        for (com.midokura.midonet.cluster.data.TunnelZone<?, ?> zoneData :
                tunnelZoneDataList) {
            TunnelZone zone = TunnelZoneFactory.createTunnelZone(zoneData);
            zone.setBaseUri(uriInfo.getBaseUri());
            tunnelZones.add(zone);
        }
        return tunnelZones;
    }

    @GET
    @RolesAllowed({AuthRole.ADMIN})
    @Path("{id}")
    @Produces({VendorMediaType.APPLICATION_TUNNEL_ZONE_JSON,
                  MediaType.APPLICATION_JSON})
    public TunnelZone get(@PathParam("id") UUID id)
        throws StateAccessException {

        com.midokura.midonet.cluster.data.TunnelZone zoneData =
                dataClient.tunnelZonesGet(id);
        if (zoneData == null) {
            throw new NotFoundHttpException();
        }

        TunnelZone zone = TunnelZoneFactory.createTunnelZone(zoneData);
        zone.setBaseUri(uriInfo.getBaseUri());

        return zone;
    }

    @DELETE
    @RolesAllowed({AuthRole.ADMIN})
    @Path("{id}")
    public void delete(@PathParam("id") UUID id)
            throws StateAccessException {

        com.midokura.midonet.cluster.data.TunnelZone zoneData =
                dataClient.tunnelZonesGet(id);
        if (zoneData == null) {
            return;
        }

        dataClient.tunnelZonesDelete(id);
    }

    @POST
    @RolesAllowed({ AuthRole.ADMIN })
    @Consumes({ VendorMediaType.APPLICATION_TUNNEL_ZONE_JSON,
            MediaType.APPLICATION_JSON })
    public Response create(TunnelZone tunnelZone)
            throws StateAccessException {

        Set<ConstraintViolation<TunnelZone>> violations = validator.validate(
                tunnelZone, TunnelZone.TunnelZoneCreateGroupSequence.class);
        if (!violations.isEmpty()) {
            throw new BadRequestHttpException(violations);
        }

        UUID id = dataClient.tunnelZonesCreate(tunnelZone.toData());
        return Response.created(
                ResourceUriBuilder.getTunnelZone(uriInfo.getBaseUri(), id))
                .build();
    }

    @PUT
    @RolesAllowed({ AuthRole.ADMIN })
    @Consumes({ VendorMediaType.APPLICATION_TUNNEL_ZONE_JSON,
            MediaType.APPLICATION_JSON })
    @Path("{id}")
    public void update(@PathParam("id") UUID id, TunnelZone tunnelZone)
            throws StateAccessException {

        tunnelZone.setId(id);

        Set<ConstraintViolation<TunnelZone>> violations = validator.validate(
                tunnelZone, TunnelZone.TunnelZoneUpdateGroupSequence.class);
        if (!violations.isEmpty()) {
            throw new BadRequestHttpException(violations);
        }

        dataClient.tunnelZonesUpdate(tunnelZone.toData());
    }

    @Path("/{id}" + ResourceUriBuilder.HOSTS)
    public TunnelZoneHostResource getTunnelZoneHostResource(
            @PathParam("id") UUID id) {
        return factory.getTunnelZoneHostResource(id);
    }
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/TunnelZoneResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.UUID
import scala.collection.JavaConverters._
import javax.ws.rs._
import javax.ws.rs.core.MediaType.APPLICATION_JSON

import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.models.Topology
import org.midonet.cluster.rest_api.ConflictHttpException
import org.midonet.cluster.rest_api.annotation._
import org.midonet.cluster.rest_api.models.{Vtep, TunnelZone}
import org.midonet.cluster.rest_api.validation.MessageProperty.{UNIQUE_TUNNEL_ZONE_NAME_TYPE, getMessage}
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.ResourceContext
import org.midonet.cluster.util.UUIDUtil

@ApiResource(version = 1)
@Path("tunnel_zones")
@RequestScoped
@AllowGet(Array(APPLICATION_TUNNEL_ZONE_JSON,
                APPLICATION_JSON))
@AllowList(Array(APPLICATION_TUNNEL_ZONE_COLLECTION_JSON,
                 APPLICATION_JSON))
@AllowCreate(Array(APPLICATION_TUNNEL_ZONE_JSON,
                   APPLICATION_JSON))
@AllowUpdate(Array(APPLICATION_TUNNEL_ZONE_JSON,
                   APPLICATION_JSON))
@AllowDelete
class TunnelZoneResource @Inject()(resContext: ResourceContext)
    extends MidonetResource[TunnelZone](resContext) {

    @Path("{id}/hosts")
    def hosts(@PathParam("id") id: UUID) = {
        new TunnelZoneHostResource(id, resContext)
    }

    protected override def createFilter(tz: TunnelZone, tx: ResourceTransaction)
    : Unit = {
        throwIfTunnelZoneNameUsed(tz)
        tx.create(tz)
    }

    protected override def updateFilter(to: TunnelZone, from: TunnelZone,
                                        tx: ResourceTransaction): Unit = {
        throwIfTunnelZoneNameUsed(to)
        to.update(from)
        tx.update(to)
    }

    /** Make sure that tunnel zone host entries in this tunnel zone have the
      * tunnel zone id and base uri populated.
      */
    private def fillTzDetails(tz: TunnelZone): TunnelZone = {
        if (tz.tzHosts != null) {
            tz.tzHosts.asScala.foreach { tzh =>
                tzh.tunnelZoneId = tz.id
                tzh.setBaseUri(uriInfo.getBaseUri)
            }
        }
        tz
    }

    override protected def getFilter(tz: TunnelZone): TunnelZone = {
        fillTzDetails(tz)
        tz
    }

    override protected def listFilter(list: Seq[TunnelZone]): Seq[TunnelZone] = {
        list.foreach(fillTzDetails)
        list
    }

    protected override def deleteFilter(id: String, tx: ResourceTransaction)
    : Unit = {
        val vteps = tx.list(classOf[Vtep])
        val matchingVteps = vteps.filter { v => v.tunnelZoneId.toString == id }
        if (matchingVteps.nonEmpty) {
            val ids = matchingVteps.map { v => v.id }
            throw new ConflictHttpException("Tunnel Zone is used by VTEP(s), " +
                                            "please delete the following " +
                                            "VTEPs before deleting this " +
                                            "tunnel zone: " + ids)
        }
        tx.delete(classOf[TunnelZone], id)
    }

    private def throwIfTunnelZoneNameUsed(tz: TunnelZone): Unit = {
        val store = resContext.backend.store
        val nameCollision = store.getAll(classOf[Topology.TunnelZone])
                                 .getOrThrow.find ( _.getName == tz.name )
        if (nameCollision.nonEmpty) {
            throw new ConflictHttpException(
                getMessage(UNIQUE_TUNNEL_ZONE_NAME_TYPE))
        }
    }

}
